Feature Refinement to Improve High Resolution Image Inpainting

ankuPRK opened the pull request:
We are a team of researchers at Geomagical Labs ([geomagical.com](https://www.geomagical.com/)), a subsidiary of IKEA.  We work on pioneering Mixed Reality apps which allow customers to scan photorealistic models of their indoor spaces and re-imagine them with virtual furniture.

In this PR we propose an additional refinement step for LaMa to improve high-resolution inpainting results. We observed that when inpainting large regions at high resolution, LaMa struggles at structure completion. However, at low resolutions, LaMa can infill the same missing region much better. To address this we added an additional refinement step that uses the structure from low resolution predictions to guide higher resolution predictions.

Our approach can work on any inpainting network, and does **not require any additional training or network modification.**

## How to run refinement

To run refinement, simply pass `refine=True` in the evaluation step as:
```
    python3 bin/predict.py refine=True model.path=$(pwd)/big-lama indir=$(pwd)/LaMa_test_images outdir=$(pwd)/output
```

## Evaluation

Here's a few example comparisons, with each triplet showing the masked image, inpainting with LaMa, and inpainting with LaMa using refinement:
![image](https://user-images.githubusercontent.com/13609663/165561990-b397dd22-7edf-4c61-bd78-2ccab6028af4.png)

Comparison of unrefined and refined images on all test images (kindly shared by you) is available here: https://drive.google.com/drive/folders/15LEa9k_7-dUKb2CPUDuw7e6Zk28KCtzz?usp=sharing

We also performed some numerical evaluation on 1024x1024 size images sampled from [[1](#unsplash)], using the thin, medium, and thick masks. Results indicate that LaMa+refinement outperforms all the recent inpainting baselines on high resultion inpainting:

<a name="tab1">

| Method | FID (thin) | LPIPS (thin) | FID (medium) | LPIPS (medium) | FID (thick) | LPIPS (thick) |
| :---         |     ---:      |          ---: |     ---:      |          ---: |     ---:      |          ---: |
| AOTGAN [[3](#aotgan)]  | 17.387 | 0.133 | 34.667 | 0.144 | 54.015 | 0.184 |
| LatentDiffusion [[4](#latentdiffusion)]  | 18.505 | 0.141 | 31.445 | 0.149 | 38.743 | 0.172 |
| MAT [[6](#mat)]  | 16.284 | 0.137 | 27.829 | 0.135 | 38.120 | 0.157 |
| ZITS [[5](#zits)]  | 15.696 | 0.125 | 23.500 | 0.121 | 31.777 | 0.140 |
| LaMa-Fourier [[2](#lama)]  | 14.780 | 0.124 | 22.584 | 0.120 | 29.351 | 0.140 |
| Big-LaMa [[2](#lama)]     | **13.143** | 0.114 | 21.169 | 0.116 | 29.022 | 0.140 |
| Big-LaMa+refinement (ours)     | 13.193 | **0.112** | **19.864** | **0.115** | **26.401** | **0.135** |

Table 1. Performance comparison of various recent inpainting approaches on 1k 1024x1024 size images 

</a>


## Video


We have also created a video to explain the technical details of our approach: 
https://www.youtube.com/watch?v=gEukhOheWgE

## References

<a id="unsplash">[1]</a>
Unsplash Dataset. https://unsplash.com/data, 2020

<a id="lama">[2]</a> 
Suvorov, R., Logacheva, E., Mashikhin, A., Remizova, A., Ashukha, A., Silvestrov, A., Kong, N., Goka, H., Park, K. and Lempitsky, V., 2022. Resolution-robust Large Mask Inpainting with Fourier Convolutions. In Proceedings of the IEEE/CVF Winter Conference on Applications of Computer Vision (pp. 2149-2159)

<a id="aotgan">[3]</a> 
Zeng, Y., Fu, J., Chao, H. and Guo, B., 2022. Aggregated contextual transformations for high-resolution image inpainting. IEEE Transactions on Visualization and Computer Graphics.

<a id="latentdiffusion">[4]</a> 
Rombach, R., Blattmann, A., Lorenz, D., Esser, P. and Ommer, B., 2021. High-Resolution Image Synthesis with Latent Diffusion Models. arXiv preprint arXiv:2112.10752.

<a id="zits">[5]</a> 
Dong, Q., Cao, C. and Fu, Y., 2022. Incremental Transformer Structure Enhanced Image Inpainting with Masking Positional Encoding. arXiv preprint arXiv:2203.00867.

<a id="mat">[6]</a> 
Li, W., Lin, Z., Zhou, K., Qi, L., Wang, Y. and Jia, J., 2022. MAT: Mask-Aware Transformer for Large Hole Image Inpainting. arXiv preprint arXiv:2203.15270.

windj007 commented:
Wow! That's extremely cool! Can't wait to try!

senya-ashukha commented:
Wow folks! that is really impressive!

windj007 commented:
Just curious, have you tried your approach with other methods, e.g. MAT or ZITS? Is the improvement for them is like that for LaMa? I'm also amazed by the fact that such an optimization technique does not introduce high-frequency artifacts... How do you think, why is it so?

ankuPRK commented:
Hi, we haven't tried it with other methods due to our limited bandwidth. We started with LaMa, and then during evaluation we found that without refinement, LaMa still generalizes better than the newer methods on high resolution inpainting. So we didn't have a very strong motivation to try this on the other methods. We'd love to see how it works on the other methods though.

When applying the refinement, we're basically asking the network to find a high-resolution featuremap that produces an output that, when downscaled, looks like the low-resolution output. We hypothesize that when this featuremap is found, it contains high-level encoded information learned from training about the contents in that region. For example, optimization may adjust a feature that was "kind-of-brick-like" to become "very-red-brick-like." The optimized high level features then gets decoded into low and high frequency brick-like textures. 

So, it's possible that these latent features of the downscaler control low and high frequency details jointly. We also tried to optimize the latent features of the upscaler, but it didn't work well, and produced cloud-like artifacts. So it probably has something to do with large and overlapping receptive fields of pixels of the featuremap.

windj007 commented:
Thank you for the clarification!

> that without refinement, LaMa still generalizes better 

Yeah, but in lower resolution these methods are stronger than LaMa - so they might probably benefit more from your method.
Maybe the difference between effects of your method on different architectures would highlight the inherent inductive biases in them - and help build a better new archtecture.

> these latent features of the downscaler control low and high frequency details jointly.

Yes, lo-freq and high-freq details do not seem to be disentangled in the features.

windj007 commented on bin/predict.py:
```diff
@@ -56,34 +56,40 @@ def main(predict_config: OmegaConf):
+                    batch['mask'] = (batch['mask'] > 0) * 1
+                    batch = model(batch)                    
+                    cur_res = batch[predict_config.out_key][0].permute(1, 2, 0).detach().cpu().numpy()
+                    unpad_to_size = batch.get('unpad_to_size', None)
```
L84-87 should be outside if-else - they need to be executed regardless refinement

ankuPRK replied:
L84-87 automatically get addressed in the refiner. Refinement is done on unpadded images so it compulsorily requires the `unpad_to_size` argument, because if we take padded input, multiscale pyramid might end up doing padding on the top of padding (when downscaling, to ensure modulo 8 is 0) which gets complicated.

Would you prefer to enable refinement even if `unpad_to_size` is `None`?

ankuPRK replied:
L84-87 already get addressed inside the refiner. Refiner works on unpadded images (it does the necesssary padding internally and then unpads the output appropriately). We can:
* add an assertion to check `unpad_to_size` is not None
* enable refiner to just run on padded image, if `unpad_to_size` is `None`.

ankuPRK replied:
We can let the refiner get padded input. But refiner still needs some padding in place. Because - 

Suppose your input image is a square of size 1000. Then the original image isn't padded because `1000%8==0`, but in the refiner, once we downscale the image, it's size becomes 500, and `500%8!=0`. So we have to pad it to make it 504x504. 

So we can't get rid of lines 301 and 302 in `refinement.py`, but we can:
- let the padded image to be input to refiner, so that we take L84-87 outside the if-else.
- refiner then doesn't check `unpad_to_size` argument at all. 
- Padding would happen in the refiner to ensure downscaled image size is divisible by 8.

windj007 replied:
> refiner still needs some padding in place

I see, thank you for the clarification! Let's just leave that piece of the as is - and add a comment about "padding-unpadding is handled within refiner"

> image size is divisible by 8.

Padding size depends on depth of the generator and thus needs to be configurable

ankuPRK replied:
Gotcha, will add the comment. Yeah the padding size of the refiner is not exactly 8, but exactly equal to dataset.pad_out_to_modulo in the predict config. I'll add a comment there in the PR

windj007 commented on configs/prediction/default.yaml:
```diff
@@ -12,3 +12,13 @@ dataset:
+
+refine: False # refiner will only run if this is True
+refiner:
+  gpu_ids: 0,1 # the GPU ids of the machine to use. If only single GPU, use: "0,"
```
I'd suggest using only 0 by default - or even introduce "None" default (so refiner would rely on the parent `device` setting). That would make this work by default in more environments without any modifications by default.

ankuPRK replied:
Actually the refiner needs around 24GB GPU to process 1.8 megapixel images (~1200x1500). Since most people have two 12GB GPUs instead, we decided to split the model onto two GPUs, that's why the default config setting. 
Do you suggest to still make it None by default?

ankuPRK replied:
torch.cuda.amp isn't working because pytorch doesn't seem to support `Half` dtype for `torch.fft.rfftn`. PFA link to the relevant issues in Pytorch repo:

https://github.com/pytorch/pytorch/issues/70664
https://github.com/pytorch/pytorch/issues/71680

windj007 replied:
> I also manually verified the requires_grad for all the params of the model

Great, thank you!

> torch.cuda.amp isn't working because pytorch doesn't seem to support Half 

Sure, I've forgot that I've already tried half and failed because of that... We could wrap rfftn/irfftn with conversion to and from .float(), but I'm not sure there wouldn't be other issues..

ankuPRK replied:
Hello @windj007, sorry for coming back after 2 months! We picked up the experiments, our findings:

1) We were able to perform the optimization in mixed precision. I haven't benchmarked it quantitatively, but qualitative results look good. However, for 1024x1024 images, it only reduces the memory from `21-22GB -> 17-18GB`, so it is still not sufficient to fit on a single 12GB GPU
2) We also tried to play with checkpointing. Performing it naively throws `RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn`, which we bypass by setting [`use_reentrant=False`](https://pytorch.org/docs/stable/checkpoint.html). However, this setting has some memory leak problem, which causes the GPU consumption to increase at each training loop, eventually leading to OOM error. We plan to raise this issue on the Pytorch repo.

FBehrad replied:
Can you please share your code for mixed-precision?

ankuPRK replied:
Sure, it's in `amp_float16` branch of our fork:

[https://github.com/geomagical/lama-with-refiner/tree/amp_float16](https://github.com/geomagical/lama-with-refiner/tree/amp_float16)

You can get this code by:
```
git clone git@github.com:geomagical/lama-with-refiner.git
git checkout amp_float16
```

Also, I've changed the config file of the refiner to run on a single GPU. But yeah feel free to play around with config parameters or anything :)

FBehrad replied:
Thank you so much 🙏

ankuPRK commented:
Thanks for your feedback, appreciate it! We are very much interested in addressing your concerns until you're comfortable enough to merge this PR into your code. The usage complication is primarily because we try to fit the refinement on multiple devices. Refinement requires at least 24GB GPU to run on images of sizes like 1024x1024.

ankuPRK commented on configs/prediction/default.yaml:
```diff
@@ -12,3 +12,13 @@ dataset:
+refine: False # refiner will only run if this is True
+refiner:
+  gpu_ids: 0,1 # the GPU ids of the machine to use. If only single GPU, use: "0,"
+  modulo: ${dataset.pad_out_to_modulo}
```
@windj007 refiner padding is defined here

senya-ashukha commented:
<img width="1117" alt="Screenshot 2022-07-23 at 21 39 24" src="https://user-images.githubusercontent.com/10505555/180616619-89aa9732-5e53-4641-94c4-7c29f1dded49.png">

Hi folks! Your results are amazing, so I've added them to the main page! Fee free to share twitter post and paper!

ankuPRK commented:
@senya-ashukha that works too! :)

ankuPRK commented:
Hi guys, 

Thanks for your shout out on the main README. Let us know if you would like us to close this PR.

senya-ashukha commented:
Hi @ankuPRK, @windj007 is on long vacation. I've merged it, but @windj007 may discuss it further as he returns.

ankuPRK commented:
Sounds good, thanks a lot! Happy to follow up on any further review comments regarding code/formatting/ideas or anything :)

mhashas commented:
is there anyway i can make this work on a gpu of 11GB? or run it on CPU?

ankuPRK commented:
@mhashas You can try using our mixed precision branch: https://github.com/saic-mdal/lama/pull/112#discussion_r935881555

We can probably have a CPU version, but it will be super slow. If you are okay with a slow CPU version, I think replacing the device with torch.device('cpu') in predict.py and refinement.py should work.

202112213501021 commented:
Can I separate out the Feature Refinement to Improve High Resolution Image Inpainting technique， just for a single image?

hjj-lmx commented:
Do you have the complete branch code?

bsbacon0966 commented:
I would like to ask the original author and users who have experience with Feature Refinement for guidance.

Currently, I am testing with the dataset provided and obtained the results shown in the image below. I would like to ask which step I might have done incorrectly.

The command I used is:

python ./bin/predict.py refine=True model.path=C:\Users\user\Desktop\lama_ori\bin\big-lama indir=C:\Users\user\Desktop\lama_ori\LaMa_test outdir=output

I was able to generate results even without refine, and adding refine also works, but the final output appears as shown in the image below.

![image](https://github.com/user-attachments/assets/b1e5f3b7-a7b6-4379-a6c3-bb4c2213af6d)

I would greatly appreciate any advice. Thank you!
